metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,619 @@
+                     icons.sync + " "
+                   else "") + getSignatureFromHover(hover).getOrElse("")
+                )
+
+                val isLocal = symbol.isLocal
+
+                val searchLocal =
+                  isLocal || source.isDependencySource(workspace) ||
+                    buildTargets.inverseSources(source).isEmpty
+
+                item.setData(
+                  CallHierarchyItemInfo(symbols, visited, isLocal, searchLocal)
+                )
+
+                item
+              })
+          )
+      case _ => Future.successful(None)
+    }
+  }
+
+  /**
+   * Prepare call hierarchy request by returning a call hierarchy item, resolved for the given text document position.
+   */
+  def prepare(params: CallHierarchyPrepareParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[List[CallHierarchyItem]] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+    semanticdbs.textDocument(source).documentIncludingStale match {
+      case Some(doc) =>
+        val results: List[ResolvedSymbolOccurrence] =
+          definition.positionOccurrences(source, params.getPosition, doc)
+        Future
+          .sequence(results.flatMap { result =>
+            for {
+              occurence <- result.occurrence
+              if occurence.role.isDefinition
+              range <- occurence.range
+              tree <- trees.findLastEnclosingAt(source, range.toLSP.getStart)
+              (definition, _) <- getSpecifiedOrFindDefinition(Some(tree))
+            } yield buildCallHierarchyItem(
+              source,
+              doc,
+              occurence,
+              definition.pos.toLSP,
+              Array(occurence.symbol),
+              token,
+            )
+          })
+          .map(_.flatten)
+      case None =>
+        Future.successful(Nil)
+    }
+  }
+
+  private def findIncomingCalls(
+      source: AbsolutePath,
+      doc: TextDocument,
+      root: Tree,
+      info: CallHierarchyItemInfo,
+  ): List[FindIncomingCallsResult] = {
+    def search(
+        tree: Tree,
+        parent: Option[Name],
+        parentRange: Option[l.Range],
+    ): List[FindIncomingCallsResult] = {
+      def default =
+        tree.children.flatMap(child => search(child, parent, parentRange))
+
+      def searchVal(valRange: l.Range, pat: Pat, rhs: Term) =
+        extractNameAndPathsFromPat(pat).flatMap { case (name, indices) =>
+          traverseTreeWithIndices(rhs, indices.toList)
+            .map(tree => search(tree, Some(name), Some(valRange)))
+            .getOrElse(default)
+        }
+
+      tree match {
+        case name: Name
+            if !isTypeDeclaration(name) && definition
+              .positionOccurrences(source, name.pos.toLSP.getEnd, doc)
+              .exists(
+                _.occurrence.exists(occ =>
+                  occ.role.isReference && info.symbols.contains(occ.symbol)
+                )
+              ) =>
```

Comment:
Same thing here with some of the guards even if they could be extracted out into a function this whole thing would read easier.